<div style="border:solid green 3px; padding: 20px">
	
<b>Привет, Александр! :)</b>
	
Меня зовут Сороколетов Илья и я буду проверять твой проект. 
	
Предлагаю общаться на «**ты**», но если тебе будет комфортнее общаться на «**вы**», то сообщи об этом в комментариях)
	
	
Для твоего удобства, я буду выделять свои комментарии следующим образом:
	
<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Самые важные замечания. Они указывают на ключевые моменты, которые влияют на конечный результат проекта. </div>
	
<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Советы или замечания, которые помогут сделать твою работу лучше, но необязательны к выполнению.
	
</div>
	
<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>
	
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения, например вот так:
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
</div>
	
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
	
Обязательно задавай вопросы если они возникнут, а я перехожу к проверке)

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv') #импортируем pandas в таблицу 'data'
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод**
вывел таблицу на экран для изучения как видно на первый взгляд, есть значения минусовые, вещественные, используется разный регистр. Требуется предобработка для приведения data frame в подходящий для анализа вид. 

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Тут необходимо применить метод info() и написать в каких столбцах есть пропуски, где нужно изменить тип данных, а также описать другие аномалии
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
1) вывел таблицу на экран для изучения
как видно на первый взгляд, есть значения минусовые, вещественные, используется разный регистр. Требуется предобработка для приведения data frame в подходящий для анализа вид.
2) использовал метод info() где увидел 2 столбца с пропусками:  'days_employed' и 'total_income'.в каждом по 2174 пропуска. 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b> 
<br>Всё верно! Целью первичного анализа данных является выявление проблем с данными, с которыми нужно будет поработать. Чем больше всего мы найдём в этом шаге, тем лучше сможем построить дальнейший план работы и понять с чем нам стоит работать, а на что можно закрыть глаза.
</div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера v2:</b> 
<br>Что можно обнаружить ещё при первичном анализе данных:    
<br>- столбец children - выбросы (20 детей)
<br>- столбец days_employed - пропуски, выбросы
<br>- столбец dob_years - нулевые значения
<br>- столбец gender - содержит артефакт
<br>- столбец total_income - пропуски
<br>- столбец purpose - много аналогичных формулировок, которые можно категоризировать
<br>Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.
<br>Во время первичного анализа данных, часто может помочь метод describe(). Он покажет тебе min/mean/max значения в каждом столбце и многое другое)
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

In [2]:
#display(data.columns) # проверям нарушения в названиях стоблцов
#data.info()
#data.isna().sum() #методом 'isna()' и 'sum()' определил
#кол-во пропусков в столбцах 'days_employed' и 'total_income' 
data['days_employed'] = data['days_employed'].fillna(0) #убираем пропуски методом 'fillna()'
data_income_type_sum = data.groupby('income_type')['total_income'].median()
print(data_income_type_sum)
data.sort_values(by='total_income') 
data['total_income'] = data['total_income'].fillna(data['total_income'].median())
data['total_income'] = data['total_income'].astype('int') 
data.isna().sum() 
data


income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64


children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21520        224791             операции с жильем  
21521        155999          сделка с автомобилем  
21522         89672                  недвижимость  
21523        244093  на покупку своего автомобиля  
21524         82047         на покупку автомобиля  

[21525 rows x 12 columns]

**Вывод**

Как правило в данных возможны пропуски в связи с человеческим фактором (возможно пользователь не указал какие-либо данные о себе). Пропуски оказались в стоблбцах 'days_employed' и 'total_income'. Для дальнейшей работы с data frame необходимо его привести в однородное состояние и ликвидировать все пропуски.

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Насколько вероятно, что клиент вообще не имеет дохода? В случае со столбцом days_employed так можно поступить, потому что он не участвует в дальнейшем исследовании, но вот столбец total_income нам нужен для ответа на вопросы в конце. Подумай, чем оптимально можно было заменить значение в этом столбце и попробуй реализовать его. Также, обрати внимание на то, что у нас есть разные типы занятости, следовательно найди ЕЁ для каждого типа и используй для замены)
</div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Так как столбец days_employed нам не нужен, то делать с ним что-то не обязательно, но я расскажу как его можно было бы обработать более оптимально:
<br>- отрицательные значения взять по модулю
<br>- пропуски заполнить медианами по возрасту
<br>- аномально высокие значения разделить на 24 и тогда получится вполне себе адекватный стаж (в реальной задаче, прежде чем такое делать, стоит уточнить могла ли быть такая ошибка при выгрузке/заполнении данных)    
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
1. сгруппировал отдельные типы занятости по столбцу 'income_type'. Вывел медианные ЗП\доходы
2. определил общую медиану среди всех по доходу и ею заменил пропуски в 'total_income'
3. сразу изменил тип данных на 'int'.
    я пока не понял как проверить пропуски в конкретных типах занятости и как потом именно в них заменить пропуски на их медиану... (?)

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b>
<br>Это можно сделать черезе transform()
    
    df_transformed = df.groupby('income_type')['total_income'].transform('median')
    df['total_income'] = df['total_income'].fillna(df_transformed)
</div>

### Замена типа данных

In [3]:
data['days_employed'] = data['days_employed'] / 365
data['days_employed'] = data['days_employed'].astype('int') # меняем тип с float на int
#data['total_income'] = data['total_income'].astype('int')
data['days_employed'] = data['days_employed'].abs() # приводим все отрицательные значения к положительным
display(data.head(10))
data.info()

children  days_employed  dob_years education  education_id  \
0         1             23         42    высшее             0   
1         1             11         36   среднее             1   
2         0             15         33   Среднее             1   
3         3             11         32   среднее             1   
4         0            932         53   среднее             1   
5         0              2         27    высшее             0   
6         0              7         43    высшее             0   
7         0              0         50   СРЕДНЕЕ             1   
8         2             18         35    ВЫСШЕЕ             0   
9         0              5         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

1) перевел значения 'float' в 'int' столбцов 'days_employed' и 'total_income' методом astype c аргументом 'int';
2) функцией 'abs()' обратил минусовые значения в столбце 'days_employed' в положительные;
2) проверил через 'info()'и на всякий случай сам data frame посмотрел. все числовые значения целые.
3) бростается в глаза невероятно большое значение в 5 строке столбца 'days_employed' женщина пенсионерка со стажем 340266 дней, что составляет 1350 лет. Больше похоже на ошибку.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>Всё верно!
</div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Что можно сделать ещё для удобства: 
<br>- разделить количество дней стажа на 365, получив таким образом стаж в годах
<br>- разделить ежемесячный доход на 1000, получив значения в тысячах. В данном случае точность до сотен рублей нам не нужна, а значения будут выглядеть аккуратнее
</div>

### Обработка дубликатов

In [4]:
data['education'] = data['education'].str.lower() #приводим к нижнему регистру все значения в столбце 'education' 
data['gender'] = data['gender'].str.lower() # тоже самое в столбце 'gender'
#data.duplicated().sum()
i = data[data.duplicated() == True] #проверяем наличие дубликатов True\False
#i
data = data.drop_duplicates() # удаляем дубликаты
data.duplicated()






0        False
1        False
2        False
3        False
4        False
         ...  
21520    False
21521    False
21522    False
21523    False
21524    False
Length: 21454, dtype: bool

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>Шикарно! Избавились от дубликатов
</div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>В данном случае ты перевёл значения в нижний регистр, чтобы это проверить тебе нужно обязательно вывести таблицу, чтобы убедиться что всё сработало как надо)
</div>

**Вывод**

Дубликаты присутсвовали в data frame во-первых по причине отсутствия единого регистра, во вторых по причине не верно занесенных данных
1) привел все значения столбцов 'education' и 'gender' к нижнему регистру методом str.lower()
2) получил общее кол-во дубликатов в data 2-мя методами - duplicated().sum()
3) выделил все строки с дубликатами и сохранил в перменной i
4) дропнул 71 дубликат
5) проверил, напечатав таблицу без дубликатов. строк стало на 71 меньше

### Лемматизация

In [5]:
from pymystem3 import Mystem
m = Mystem()
real_purpose = data['purpose'].unique()
real_purpose = ' '.join(real_purpose)
lemmas = m.lemmatize(real_purpose)
from collections import Counter
display(Counter(lemmas))

Counter({'покупка': 10,
         ' ': 96,
         'жилье': 7,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 2,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1,
         '\n': 1})

**Вывод**

1) импортировал библиотеку pymystem3
2) Для наглядности вывел все преобразования переменной real_purpose в отдельных строках
3) импортировал специальный контейнер Counter из модуля collections для подсчета числа лематизированных слов
4) при помощи лематизации мне стало видно определенное кол-во целей, так как кол-во форм(падежей) слов существенно сократилось. Так удобнее будет кагоризировать в дальнейшем 

### Категоризация данных

In [6]:
from nltk.stem import SnowballStemmer # импортируем стеммер для русского языка
russian_stemmer = SnowballStemmer('russian')  

words = ['недвижимость', 'автомобиль', 'образование', 'свадьба', 'жилье',]
for w in words: 
    display('Исходное слово - ' + w + ', после стемминга - ' + russian_stemmer.stem(w))#для каждого
    #элемента списка напечатаем исходное слово и его основу, выявленную стеммингом

def purpose_groups(purpose): # создаем функцию для категоризации по целям
    if ('недвижим' in purpose) or ('жил' in purpose):
            return 'Недвижимость'
    if 'автомобил' in purpose:
            return 'Автомобиль'
    if 'образован' in purpose:
            return 'Образование'
    if 'свадьб' in purpose:
            return 'Свадьба'       
    return 'Другое'    
 
data['purpose_groups'] = data['purpose'].apply(purpose_groups) #создаем новый столбец
#с категориями по целям

def children_groups(children): #создаем функцию для категоризации по наличию и кол-ву детей 
    if children > 0:
        return 'есть дети'
    return 'бездетные'
data['children_groups'] = data['children'].apply(children_groups) # создаем новый столбец с категориями по детям
#display(data.head(10))

display(data['total_income'].sort_values(ascending=False)) # отсортировал доходы по убыванию, чтобы определить на какое кол-во групп делить
def income_groups(total_income): #создаем функцию для категоризации по уровню дохода
    if (total_income >= 0) and (total_income < 500000):
        return 'низкий доход'
    if (total_income > 500000) and (total_income < 1000000) :
        return 'средний доход'
    if (total_income > 1000000) and (total_income < 2000000) :    
        return 'высокий доход'
    return 'олигарх'
        
data['income_groups'] = data['total_income'].apply(income_groups) # создаем новый столбец с категориями по уровню дохода
display(data.head(10))
        






'Исходное слово - недвижимость, после стемминга - недвижим'

'Исходное слово - автомобиль, после стемминга - автомобил'

'Исходное слово - образование, после стемминга - образован'

'Исходное слово - свадьба, после стемминга - свадьб'

'Исходное слово - жилье, после стемминга - жил'

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


12412    2265604
19606    2200852
9169     1726276
20809    1715018
17178    1711309
          ...   
14276      21895
1598       21695
16174      21367
13006      21205
14585      20667
Name: total_income, Length: 21454, dtype: int64

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


children  days_employed  dob_years education  education_id  \
0         1             23         42    высшее             0   
1         1             11         36   среднее             1   
2         0             15         33   среднее             1   
3         3             11         32   среднее             1   
4         0            932         53   среднее             1   
5         0              2         27    высшее             0   
6         0              7         43    высшее             0   
7         0              0         50   среднее             1   
8         2             18         35    высшее             0   
9         0              5         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      f   сотрудник     0        253875   
1   женат / замужем                 0      f   сотрудник     0        112080   
2   женат / замужем                 0      m   сотрудник     0        145885   
3   женат / замужем                 0      m   сотрудник     0        267628   
4  гражданский брак                 1      f   пенсионер     0        158616   
5  гражданский брак                 1      m   компаньон     0        255763   
6   женат / замужем                 0      f   компаньон     0        240525   
7   женат / замужем                 0      m   сотрудник     0        135823   
8  гражданский брак                 1      f   сотрудник     0         95856   
9   женат / замужем                 0      m   сотрудник     0        144425   

                      purpose purpose_groups children_groups income_groups  
0               покупка жилья   Недвижимость       есть дети  низкий доход  
1     приобретение автомобиля     Автомобиль       есть дети  низкий доход  
2               покупка жилья   Недвижимость       бездетные  низкий доход  
3  дополнительное образование    Образование       есть дети  низкий доход  
4             сыграть свадьбу        Свадьба       бездетные  низкий доход  
5               покупка жилья   Недвижимость       бездетные  низкий доход  
6           операции с жильем   Недвижимость       бездетные  низкий доход  
7                 образование    Образование       бездетные  низкий доход  
8       на проведение свадьбы        Свадьба       есть дети  низкий доход  
9     покупка жилья для семьи   Недвижимость       бездетные  низкий доход

**Вывод**
1. категоризация помогла уменьшить кол-во уникальных значений по каждому столбцу, посредством присоединения новых столбцов. Получилось создать необходмые Series, применить метод apply и соединить их вновь в data frame. Принцип *split-apply-combine.*
2. анализ группы по кредитным целям показал: наибольшее кол-во заявок приходится на Недвижимость =    10811, а наименьшее для Свадеб и создания будущих семей, всего 2324.
3. посмотрев на наличие и кол-во детей можно сделать однозначный вывод о том, что в нашей базе превалирует числоБездетных = 14138, а многодетных как и следовало ожидать - меньшинство = 2508
4. Семейных пар оказалось больше с заявками на кредит, чем всех остальных (всего до группировки было 4 значения в столбце, что было очень не удобно для анализа)
5. группировка по уровню дохода показала, что за кредитом обращаются в основном люди с низким годовым доходом = более 23000, все остальные сильно позади, при этом люди с высоким уровнем и "олигархи" так же всписке тех, кто кредиты берут.
6. определить зависимость наличия долга от новых столбцов и их категорий теперь возможно.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>Цели кредита категоризированы правильно! 👍
</div>

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Вопрос в шаге 3 звучит так: "Есть ли зависимость между наличием детей и возвратом кредита в срок?". Здесь всех нужно разделить на 2 категории - есть дети или нет
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
1. принцип категоризации изменил. В фунции теперь только одно условие: есть дети или нет (при не выполнении


<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>У нас уже есть категоризация по семейному положению и если ты делаешь ещё одну, то наш анализ становится более общим.
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
убрал функцию, которая категоризирует по семейному положению

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>Так можно выделить категории уровня дохода, но в реальной задаче лучше этот момент уточнить у заказчика или коллег. Желательно выделить 4-5 категорий, чтобы можно было сделать более точный анализ
</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [7]:
data_children_count = data.groupby('children_groups')['debt'].count() # создадим series где видно разделение семей по кол-ву детей 
display(data_children_count)
data_children_sum = data.groupby('children_groups')['debt'].sum() # создадим series где видно кол-во должников относительно каждой группы 
display(data_children_sum)
data_children_percent = (data_children_sum/data_children_count).sort_values()
display(data_children_percent)


children_groups
бездетные    14138
есть дети     7316
Name: debt, dtype: int64

children_groups
бездетные    1064
есть дети     677
Name: debt, dtype: int64

children_groups
бездетные    0.075258
есть дети    0.092537
Name: debt, dtype: float64

**Вывод**

1. Бездетные люди чаще возвращают кредит и среди должников их всего **7,5%**, что говорит об их большей надежности по сравнении с остальными группами.
2. семьи с одним ребенком берут кредит в **2 раза реже**, но при этом их плохая статистика равна **9,3%** от их общего числа.
3. из представленных групп можно отдавать наибольшее предпочтение группе бездетных семей, так как у них наименьшее число иждивенцев и, следовательно, процент возратов больший чем у семей с детьми.

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Вывод у тебя правильный, но нужно поправить категоризацию
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
поправил выводы, так как категории стало 2

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b>
<br>Супер) Теперь всё верно!
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [8]:
data_family_id_count = data.groupby('family_status')['debt'].count() # создадим series где видно разделение семей по семейному положению
display(data_family_id_count.head(10))
data_family_id_sum = data.groupby('family_status')['debt'].sum() # создадим series где видно кол-во должников относительно каждой группы 
display(data_family_id_sum.head(10))
data_family_id_percent = (data_family_id_sum/data_family_id_count).sort_values()
display(data_family_id_percent)

family_status
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
гражданский брак          4151
женат / замужем          12339
Name: debt, dtype: int64

family_status
Не женат / не замужем    274
в разводе                 85
вдовец / вдова            63
гражданский брак         388
женат / замужем          931
Name: debt, dtype: int64

family_status
вдовец / вдова           0.065693
в разводе                0.071130
женат / замужем          0.075452
гражданский брак         0.093471
Не женат / не замужем    0.097509
Name: debt, dtype: float64

**Вывод**

1. разделив семейныепары на 5 групп, сразу стало видно, что официальные семейные пары берут кредит чаще, но при этом не имеют большую дисциплинированность при его возврате. **7,5%** должников.
2. не оформившие официальный брак пары, а также "одиночки" имеют массу возможностей к расставаниию, а как следствие потере возможности вернуть кредит, о чем и говорит статистика: **9,3%** и **9,8%** должников от их общей массы соотвественно. 
3. лидер по кредитной дисциплине внезапно оказалась группа **водец / вдова** с результатом **6,5%** должников, однако нужно обратить внимание на самое маленькое кол-во заявок относительно остальных групп.
3. Очевидно давать предпочтение по выдаче кредита нужно в первую очередь **водец / вдова**, людям в разводе и официальным семейным парам соответственно.

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Вывод у тебя практически правильный. После того как ты поправишь категоризацию, лидер должен смениться)
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
поправил выводы после того как обновились данные по категориям семейного положения

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера v2:</b>
<br>Замечательно! 
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [9]:
data_income_count = data.groupby('income_groups')['debt'].count() # создадим series где видно раздельное кол-во целей 
display(data_income_count.head(10))
data_income_sum = data.groupby('income_groups')['debt'].sum() # создадим series где видно кол-во должников относительно каждой группы целей
display(data_income_sum.head(10))
data_income_percent = (data_income_sum/data_income_count).sort_values()
display(data_income_percent)

income_groups
высокий доход       23
низкий доход     21232
олигарх              2
средний доход      197
Name: debt, dtype: int64

income_groups
высокий доход       1
низкий доход     1727
олигарх             1
средний доход      12
Name: debt, dtype: int64

income_groups
высокий доход    0.043478
средний доход    0.060914
низкий доход     0.081339
олигарх          0.500000
Name: debt, dtype: float64

**Вывод**

1. отсортировав группы по доходам и вычислив долю должников стало понятно что самые надежные находятся в группе высоких доходов с **4,3%** должников, а самые рискованные заемщики - люди с низкими доходами = **8,1%** должников.
2. при этом самый страшный процент должников = 50% лежит в группе "Олигарх", но там все 2 участника и эта группа не репрезентативна 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>В точку! Идём дальше)
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [10]:
data_purpose_count = data.groupby('purpose_groups')['debt'].count() # создадим series где видно раздельное кол-во целей 
display(data_purpose_count)
data_purpose_sum = data.groupby('purpose_groups')['debt'].sum() # создадим series где видно кол-во должников относительно каждой группы целей
display(data_purpose_sum)
data_purpose_percent = (data_purpose_sum/data_purpose_count).sort_values()
display(data_purpose_percent)

data_pivot = data.pivot_table(
    index=['purpose_groups', 'purpose'], columns='children_groups', values='debt', aggfunc='sum')
data_pivot

purpose_groups
Автомобиль       4306
Недвижимость    10811
Образование      4013
Свадьба          2324
Name: debt, dtype: int64

purpose_groups
Автомобиль      403
Недвижимость    782
Образование     370
Свадьба         186
Name: debt, dtype: int64

purpose_groups
Недвижимость    0.072334
Свадьба         0.080034
Образование     0.092200
Автомобиль      0.093590
Name: debt, dtype: float64

children_groups                                        бездетные  есть дети
purpose_groups purpose                                                     
Автомобиль     автомобили                                     28         16
               автомобиль                                     31         11
               на покупку автомобиля                          22         22
               на покупку подержанного автомобиля             20         16
               на покупку своего автомобиля                   30         16
               приобретение автомобиля                        23         19
               свой автомобиль                                27         21
               сделка с автомобилем                           34         16
               сделка с подержанным автомобилем               29         22
Недвижимость   жилье                                          31         15
               недвижимость                                   29         13
               операции с жильем                              30         18
               операции с коммерческой недвижимостью          34         18
               операции с недвижимостью                       33         22
               операции со своей недвижимостью                28         22
               покупка жилой недвижимости                     23         18
               покупка жилья                                  28         20
               покупка жилья для сдачи                        31         21
               покупка жилья для семьи                        29         16
               покупка коммерческой недвижимости              30         17
               покупка недвижимости                           21         22
               покупка своего жилья                           20         14
               ремонт жилью                                   23         12
               строительство жилой недвижимости               37         11
               строительство недвижимости                     27         27
               строительство собственной недвижимости         22         20
Образование    высшее образование                             24         16
               дополнительное образование                     26         12
               заняться высшим образованием                   28         15
               заняться образованием                          20         19
               образование                                    25          7
               получение высшего образования                  28         18
               получение дополнительного образования          28         23
               получение образования                          21         16
               профильное образование                         29         15
Свадьба        на проведение свадьбы                          39         25
               свадьба                                        39         25
               сыграть свадьбу                                37         21

**Вывод**

1. зависимость возврата кредита от их целей тоже есть. Например наибольшее число заявок в группе "Недвижимость" и при этом там же наименьший процент должников =**7,2%**
2. Группы "Образование" и "автомобиль" примерно равны как по общему числу заявок, так и по проценту должников = **9,2%** и **9,3%** соответственно.
3. На "Свадьбу" кредитов берут меньше всего и эта группа в числе "хороших" заемщиков с результатом **8%** должников, так как создание молодой официальной семьи как правило ведет к увеличению суммарного дохода и положительно сказывается на финансовой дисциплине заемщиков.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b>
<br>Абсолютно верно! Возможно, это связано с тем, что к покупке жилья люди относятся более ответственно, так как сумма заемных средств и срок кредита, как правило гораздо выше чем в остальных категориях, а свадебные кредиты возвращаются за счет подаренных денег
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
построил дополнительную сводную таблицу

## Шаг 4. Общий вывод

1. **Влияение семейного положения и наличие детей на возврат кредита очевидно.** 
2. **после группировки и сортировки данных можно склеить портрет почти идеального заемщика: это семейный человек, либо водец/вдова, без детей, с высоким либо средним доходом.** 
3. **Так же статистика подсказывает тот факт, что при выборе целей кредита тоже есть разница. Если "идеальный клиент" выбирает покупку недвижимости, то желание вернуть кредит выше, а процент должников ниже и составляет 7,2%.**
4. **стоит обратить особое внимание на портрет неидеального клиента, а именно - многодетный отец \ мать, не оформившие официальные отношения, с низким уровнем дохода и с целью оформить авто кредит.**
*Итоговая таблица с исследуемыми категориями клиентов находится ниже.*

<div class="alert alert-danger">
<b>❌ Итоговый комментарий ревьюера:</b> 
<br>Ты хорошо потрудился, большая часть работы сделана, но остаётся внести следующие правки:
<br>- применить метод info() для первичного анализа данных    
<br>- убрать дополнительную категоризацию по семейному положению  
<br>- правильно категоризировать наличие детей   
<br>- построить сводную таблицу используя метод pivot_table() для любого из вопросов в шаге 3. Это необходимо для отработки пройденного материала
<br>- поправить итоговые выводы    
</div>

<div class="alert alert-success">
<b>✔️ Итоговый комментарий ревьюера v2:</b>
<br>Ты проделал большую работу! У тебя шикарные выводы с хорошей аргументацией. Здорово, что ты пишешь свои рассуждения и предположения - это помогает видеть ход твоей мысли. Отдельно хочу выделить аккуратность и структурированность твоей работы. Отдельный плюс за итоговую таблицу в конце - это готовый отчет для коллег) Рад был поработать над проверкой твоего проекта! Желаю удачи в дальнейшем обучении! 😉
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

In [11]:
data_table_children = [['бездетные', '14196', '1064', '0.074951'],
       ['один ребенок', '4818', '444', '0.092154'],
       ['многодетные', '2511', '233', '0.092792']]       
columns = ['категория заемщика', 'общее кол-во', 'должники', 'процент должников']

research_children_result = pd.DataFrame(data = data_table_children, columns =
columns)
display(research_children_result)

data_table_family_id = [['женат / замужем', '12380', '931', '0.075202'],
       ['не женат / не замужем', '9145', '810', '0.088573']]    
columns = ['категория заемщика', 'общее кол-во', 'должники', 'процент должников']

research_family_id_result = pd.DataFrame(data = data_table_family_id, columns =
columns)
display(research_family_id_result)

data_table_income = [['высокий доход', '23', '1', '0.043478'],
       ['средний доход', '197', '12', '0.060914'],
       ['низкий доход', '21303', '1727', '0.081068'], 
       ['олигарх', '2', '1', '0.50']]                  
columns = ['категория заемщика', 'общее кол-во', 'должники', 'процент должников']

research_income_result = pd.DataFrame(data = data_table_income, columns =
columns)
display(research_income_result)

data_table_purpose = [['недвижимость', '10840', '782', '0.072140'],
       ['свадьба', '2348', '186', '0.091994'],
       ['образование', '4022', '370', '0.091994'], 
       ['автомобиль', '4315', '403', '0.093395']]                  
columns = ['категория заемщика', 'общее кол-во', 'должники', 'процент должников']

research_purpose_result = pd.DataFrame(data = data_table_purpose, columns =
columns)
display(research_purpose_result)

категория заемщика общее кол-во должники процент должников
0          бездетные        14196     1064          0.074951
1       один ребенок         4818      444          0.092154
2        многодетные         2511      233          0.092792

категория заемщика общее кол-во должники процент должников
0        женат / замужем        12380      931          0.075202
1  не женат / не замужем         9145      810          0.088573

категория заемщика общее кол-во должники процент должников
0      высокий доход           23        1          0.043478
1      средний доход          197       12          0.060914
2       низкий доход        21303     1727          0.081068
3            олигарх            2        1              0.50

категория заемщика общее кол-во должники процент должников
0       недвижимость        10840      782          0.072140
1            свадьба         2348      186          0.091994
2        образование         4022      370          0.091994
3         автомобиль         4315      403          0.093395